File: python/leetcode/advenced-algorithm/divide-and-conquer/different-ways-to-add-parentheses.py
from typing import List


class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        if expression.isdigit():
            return [int(expression)]

        results = []
        for index, value in enumerate(expression):
            if value in "+-*":
                left = self.diffWaysToCompute(expression[:index])
                right = self.diffWaysToCompute(expression[index + 1 :])
                for l in left:
                    for r in right:
                        results.append(self.compute(l, r, value))

        return results

    def compute(self, left, right, op):
        if op == "+":
            return left + right
        elif op == "-":
            return left - right
        return left * right

File: javascript/leetcode/advanced-algorithm/divide-and-qonquer/different-ways-to-add-parentheses.js
const compute = (left, right, op) => {
  if (op === "+") {
    return left + right;
  } else if (op === "-") {
    return left - right;
  }
  return left * right;
};

const helper = (left, right, op) => {
  const results = [];
  for (let l of left) {
    for (let r of right) {
      results.push(compute(l, r, op));
    }
  }
  return results;
};
/**
 * @param {string} expression
 * @return {number[]}
 */
var diffWaysToCompute = function (expression) {
  if (!isNaN(expression)) return [Number(expression)];

  const results = [];
  for (let i = 0; i < expression.length; i++) {
    if ("+-*".includes(expression[i])) {
      const left = diffWaysToCompute(expression.slice(0, i));
      const right = diffWaysToCompute(expression.slice(i + 1));

      results.push(...helper(left, right, expression[i]));
    }
  }
  return results;
};
